Update `update_post_meta()` code where an order ID is passed to use CRUD

mattallan opened the pull request:
Fixes #142

## Description

<!--
Write a brief summary about this PR. 
- Why is this change needed? 
- What does this change do? 
- Were there other solutions you considered? 
- Why did you choose to pursue this solution? 
- Describe any trade-offs you might have had to make.
-->

When order data is moved to custom tables, updating order meta using `update_post_meta()` will no longer work. This PR updates all cases of `update_post_meta()` where an Order ID is passed - thankfully there wasn't many cases in our codebase 😅 

## How to test this PR

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
Use the testing instructions from the linked issue as a starting point.
-->

**Testing `wcs_set_objects_property()` changes**

This function is pretty well covered in unit tests, however to test nothing is broken a simple test would be to process a subscription renewal order and make sure the renewal order has the correct meta data.

1. With an active subscription, process a renewal order
2. Confirm the new renewal order has `_billing_address_1` metadata set

**Testing `save_old_paypal_meta()` changes**

This one was a little trickier to test so I just hacked some of the conditions because the only changes we cared about is the order updating post meta:
1. Edit the `save_old_paypal_meta()` function to pass if conditions using, `true ||`
2. Manually set `$subscriptions` to be `$subscriptions = [ 101823 => wcs_get_subscription( 101823 ) ];`
3. Run the following snippet:

```
add_action( 'admin_footer', function() {
	update_post_meta( 101823, '_paypal_subscription_id', 'S-fake1234' );
	update_post_meta( 101823, '_payment_method', 'paypal' );

	WCS_PayPal_Standard_Switcher::save_old_paypal_meta( 101824, [] );
} );
```

4. Confirm the $order meta is set in the DB

![image](https://user-images.githubusercontent.com/2275145/165427171-425642d1-86d8-4657-8fd8-e06601cbbefd.png)

## Product impact
<!-- What products will this PR ship in? -->

- [x] Added changelog entry (or does not apply)
- [x] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [ ] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref

Commit:
Replace update_post_meta(order_id) calls in save_old_paypal_meta()

Commit:
Update wcs_set_objects_property() to remove post meta (pre WC 3.0) calls

Commit:
Add/update PHPUnit tests for wcs_set_objects_property()

Commit:
add changelog

brucealdridge commented:
@mattallan The new unit test you added is currently failing.

Commit:
Fix failing unit tests

I copied these unit tests from the PR that introduced these changes on WC Subscriptions
and didn't mean to copy the unit tests that expected the deprecation notice.

We're not changing the wcs_get_object_property function in PR

mattallan commented:
> The new unit test you added is currently failing.

Whoops! Thanks for the heads up! I copied the unit test changes from https://github.com/woocommerce/woocommerce-subscriptions/pull/3814/files but didn't realise some of the unit test changes were for `wcs_get_object_property()` which isn't changed in this PR.

I've removed the expected deprecated unit test but kept the test additions.

@brucealdridge this should be good to review again. Thanks

mattallan commented:
One of the existing tests is failing with the unexpected calculated date comparison (off by 1 second). I'll try run the test again but this failing test shouldn't block this PR

Commit:
fix test flakiness

mattallan commented:
I just pushed up [3d1e6f0](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/3d1e6f087345ed79bdd458a6ece25e43bd31b730) which should help improve test flakiness 🤞

Commit:
Remove changelog for individual WP Post API changes

brucealdridge commented:
> I just pushed up [3d1e6f0](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/3d1e6f087345ed79bdd458a6ece25e43bd31b730) which should help improve test flakiness 🤞

This is issue #157  
I did some research on it and it looks like `assertEquals` has removed the delta parameter in the newer version of phpunit so we will need to use `assertEqualsWithDelta` instead.

mattallan commented:
> This is issue https://github.com/Automattic/woocommerce-subscriptions-core/issues/157

Oh nice! Apologies I didn't see an issue was already created for this!

I'm happy to take it off this PR and handle it separately to avoid confusion that said, I kinda have an issue with using `assertEqualsWithDelta()` where it can be avoided.

In this instance, the issue with this test sometimes failing and sometimes passing is because we're setting up a subscription's start date with `time() -1 month` and then we're calling a bunch of subscription operations (updating dates and status) and then testing that the end date is `assertEqual` to `time()`.

Depending on how slow the test environment is and how much we set as a delta (i.e. 1-2 seconds), we could still this test failing in the future unless we blow out the delta to 5-10 seconds.

The solution I went with was to use a set/stored version `time()` variable to assert expected date calculations, rather than assuming the operations made during the test will be completed within a given delta 😄

brucealdridge commented on includes/gateways/paypal/includes/class-wcs-paypal-standard-switcher.php:
```diff
@@ -155,9 +155,10 @@ public static function maybe_set_payment_method( $payment_processing_result, $or
 
-		if ( wcs_order_contains_switch( $order_id ) ) {
-			$subscriptions = wcs_get_subscriptions_for_order( $order_id, array( 'order_type' => 'switch' ) );
+		if ( wcs_is_order( $order ) && wcs_order_contains_switch( $order ) ) {
```
I'd be tempted to change this condition to do a early return and remove a level of indentation from the method 
```suggestion
		if ( ! wcs_is_order( $order ) || ! wcs_order_contains_switch( $order ) ) {
		    return;
		}
```

mattallan replied:
Fixed in 886238ca. Thanks for the suggestion!

brucealdridge commented on tests/unit/test-wcs-compatibility-functions.php:
```diff
@@ -23,12 +23,18 @@ public function test_wcs_get_objects_property() {
 		$this->assertEquals( get_post( $variation_product->get_parent_id() ), wcs_get_objects_property( $variation_product, 'post' ) );
 
-		// post_status.
+		// post_status - now deprecated as a property
```
If these are deprecated can we remove them from the unit tests?
Or move to another function and mark as deprecated?

mattallan replied:
This param isn't deprecated (copy/paste fail from original PR made to WC Subscriptions). I've removed this wording from the comment to avoid confusion: [4c72ed4](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/4c72ed41d623c4ed5f43f3980e845b24b07f0f6e)

brucealdridge commented on tests/unit/test-class-wc-subscriptions.php:
```diff
@@ -1137,13 +1137,13 @@ public function test_update_status_to_pending_canellation() {
 				try {
-
-					$start_date = gmdate( 'Y-m-d H:i:s', strtotime( '-1 month' ) );
+					$expected_end_date = gmdate( 'Y-m-d H:i:s' );
+					$start_date        = gmdate( 'Y-m-d H:i:s', strtotime( $expected_end_date . ' -1 month' ) );
```
I've tested this by adding a sleep(1); between lines 1141-1142 and 1142-1143 and in both cases they fail

mattallan replied:
I see 🙈 Putting sleeps in the test still causes the same issues that I was trying to prevent.

Because the subscription doesn't have a next payment date, the new calculated end date is being set to the current time, therefore putting sleeps between these lines causes the expected end date I saved to be in the past and an end date can't be in the past, so it still fails.

I had a solution to this by using dates in the future and setting a next payment date on the subscription (which would be the new expected end date), but I think the original purpose of this test is to make sure any end date is set as a result of updating the subscription status to pending-cancel (not validating the end date per se). My approach would slow the test down by updating more dates etc so I'll revert back to using withDelta and go back to the original code 😅

mattallan replied:
For reference, here's the updated test I had:

```
$start_time        = time();
$expected_end_time = strtotime( '+1 month', $start_time );

$subscription->update_dates( [
	'start'        => gmdate( 'Y-m-d H:i:s', $start_time ),
	'next_payment' => gmdate( 'Y-m-d H:i:s', $expected_end_time ),
] );

$subscription->update_status( 'pending-cancel' );

$this->assertEquals( $expected_end_time, $subscription->get_time( 'end' ) );
```

mattallan replied:
Fixed in [a218e45](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/a218e4549b1017bce6c82aecfccdc78d03ace860)

brucealdridge reviewed (approved):
Just a few minor nitpicks

Commit:
Use assertEqualsWithDelta to fix tests failing with off-by-one timestamp errros

Commit:
Fix copy/paste comment error

Commit:
Refactor save_old_paypal_meta to return early instead of indent whole function

mattallan commented:
Thanks for the review @brucealdridge . I've addressed the issues you've raised. Since you've approved this PR i'll merge it when everything is green :+1:

Fixes #157
^ is fixed by [a218e45](https://github.com/Automattic/woocommerce-subscriptions-core/pull/159/commits/a218e4549b1017bce6c82aecfccdc78d03ace860)
